### PR TITLE
Fix/ad policy events

### PR DESCRIPTION
**Pull request by brendanobra**
## What

This addresses 2 bugs
- Ripple Not emitting advertising.onPolicyChanged when skipRestriction value is changed
- unwrap when thunder call fails, causing panic

## Why

Why are these changes needed?

## How

How do these changes achieve the goal?
- for advertising.onPolicyChanged, an additional (missing) event key was added
- for thunder change, unwrap was replaced with return of empty value 

## Test

Manually against hardware

## Checklist

- [ X] I have self-reviewed this PR
- [ X] I have added tests that prove the feature works or the fix is effective

### CODE REVIEW

**Review by satlead — APPROVED** [review]

**Review by Vinodsathyaseelan — APPROVED** [review]

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 1]
```diff
@@ -433,3 +412,66 @@ impl RippleRPCProvider<AdvertisingImpl> for AdvertisingRPCProvider {
...
+            id: Some(1),
+            method: method_name,
+            params: Some(the_map),
+        })
+        .unwrap();
+        serde_json::to_string(&v).unwrap()
+    }
+
+    #[tokio::test]
+    pub async fn test_app_bundle_id() {
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:463]: Adding this as a starting point for one possible approach to testing these handlers.  This does not work for any functions  that are integrated (anything that uses `StorageManager` directly in this case).

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 2]
```diff
@@ -433,3 +412,66 @@ impl RippleRPCProvider<AdvertisingImpl> for AdvertisingRPCProvider {
...
+    use ripple_tdk::utils::test_utils::Mockable;
+    use serde::{Deserialize, Serialize};
+    use serde_json::json;
+
+    #[derive(Serialize, Deserialize, Clone, Debug)]
+    struct CallContextContainer {
+        pub ctx: Option<CallContext>,
+    }
+
+    fn merge(a: &mut Value, b: &Value) {
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:429]: If this is deemed useful, probably good to move it into a util of some sort

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 3]
```diff
@@ -397,14 +379,11 @@ impl AdvertisingServer for AdvertisingImpl {
...
     async fn advertising_skip_restriction(&self, _ctx: CallContext) -> RpcResult<String> {
-        let skip_restriction =
-            StorageManager::get_string(&self.state, StorageProperty::SkipRestriction).await;
-
-        let sr = match skip_restriction {
-            Err(_) => String::from("none"),
-            Ok(_) => skip_restriction.unwrap(),
-        };
-        Ok(sr)
+        Ok(
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:382]: More idiomatic and less stuff

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 4]
```diff
@@ -359,27 +346,22 @@ impl AdvertisingServer for AdvertisingImpl {
...
+        Ok(get_advertisting_policy(&self.state).await)
     }
 
     async fn advertising_on_policy_changed(
         &self,
         ctx: CallContext,
         request: ListenRequest,
     ) -> RpcResult<ListenerResponse> {
-        rpc_add_event_listener(&self.state, ctx, request, EVENT_ADVERTISING_POLICY_CHANGED).await
+        rpc_add_event_listener_with_decorator(
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:357]: This is the actual bugfix :)

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 5]
```diff
@@ -149,14 +162,7 @@ impl AppEventDecorator for AdvertisingPolicyEventDecorator {
...
     ) -> Result<Value, AppEventDecorationError> {
-        let skip_restriction =
-            StorageManager::get_string(ps, StorageProperty::SkipRestriction).await;
-
-        let sr = match skip_restriction {
-            Err(_) => String::from("none"),
-            Ok(_) => skip_restriction.unwrap(),
-        };
-        Ok(serde_json::to_value(sr).unwrap())
+        Ok(serde_json::to_value(get_advertisting_policy(ps).await)?)
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:165]: More idiomatic and less stuff

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 6]
```diff
@@ -175,14 +181,11 @@ impl AppEventDecorator for AdvertisingSetRestrictionEventDecorator {
...
     ) -> Result<Value, AppEventDecorationError> {
-        let skip_restriction =
-            StorageManager::get_string(ps, StorageProperty::SkipRestriction).await;
-
-        let sr = match skip_restriction {
-            Err(_) => String::from("none"),
-            Ok(_) => skip_restriction.unwrap(),
-        };
-        Ok(serde_json::to_value(sr).unwrap())
+        Ok(serde_json::to_value(
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:184]: More idiomatic and less stuff

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 7]
```diff
@@ -137,7 +134,23 @@ pub trait Advertising {
     #[method(name = "advertising.resetIdentifier")]
     async fn reset_identifier(&self, ctx: CallContext) -> RpcResult<()>;
 }
-
+const NONE: &str = "none";
+async fn get_advertisting_policy(platform_state: &PlatformState) -> AdvertisingPolicy {
```
Comment (brendanobra) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:138]: free function to cut down on repetition

File: core/sdk/src/api/firebolt/fb_capabilities.rs [hunk 8]
```diff
@@ -339,6 +339,21 @@ pub enum DenyReason {
     GrantProviderMissing,
     AppNotInActiveState,
 }
+impl std::fmt::Display for DenyReason {
```
Comment (brendanobra) [inline core/sdk/src/api/firebolt/fb_capabilities.rs:342]: This makes returning DenyReason much easier over jsonrpc, and is more idiomatic

File: core/main/src/firebolt/handlers/advertising_rpc.rs [hunk 9]
```diff
@@ -137,7 +134,23 @@ pub trait Advertising {
...
+    AdvertisingPolicy {
+        skip_restriction: StorageManager::get_string(
+            platform_state,
+            StorageProperty::SkipRestriction,
+        )
+        .await
+        .unwrap_or_else(|_| String::from(NONE)),
+        limit_ad_tracking: StorageManager::get_bool(
```
Comment (Vinodsathyaseelan) [inline core/main/src/firebolt/handlers/advertising_rpc.rs:146]: You may need to use the the following function to get limit_ad_tracking
`limit_ad_tracking: !PrivacyImpl::get_allow_app_content_ad_targeting(ps).await;`

**Review by Vinodsathyaseelan — CHANGES_REQUESTED** [review]
fix needed for limit_ad_tracking.

**Review by Vinodsathyaseelan — APPROVED** [review]

**Pull request merged**